numpy/core/tests/test_umath.py

```diff
@@ -267,6 +267,14 @@ def __ne__(self, other):
         a = np.array([np.nan], dtype=object)
         assert_equal(np.not_equal(a, a), [True])
 
+    def test_error_in_equal_reduce(self):
+        # gh-20929
+        # make sure np.equal.reduce raises a TypeError if an array is passed
+        # without specifying the dtype
+        a = np.array([0, 0])
+        assert_equal(np.equal.reduce(a, dtype=bool), True)
+        assert_raises(TypeError, np.equal.reduce, a)
```

Comment:
This is causing a segfault in the debug builds.

Reply:
hmmm sounds like hitting an assert, probably just a wrong one...

Reply:
Thank you for the review, @charris and @seberg! I am very new to the C++ and Python combination. Could you please help me with the segfault, or link a similar error and fix so that I can have look? I tried running the testing suite by removing the `assert_raises` test and it does pass.

Reply:
Hmmmpf, should be able to get a traceback using gdb: https://numpy.org/devdocs/dev/development_environment.html#debugging

But I cannot reproduce with a python3.9 debug version and see nothing with valgrind either.  Can you undo the last commit, because I don't know anymore what is running on CI.  At least if it is a linux debug print, we can also hook in gdb in CI (although I have to look up how to every time).
If you can reproduce a crash with `-g` locally, then running `gdb` will tell us more.

Reply:
Oh, I can reproduce, it requires a full test-suite run though... maybe something wrong related to the dispatching caching.

Reply:
OK, I think the assert is indeed incorrect and it should be something like:
```diff
diff --git a/numpy/core/src/umath/dispatching.c b/numpy/core/src/umath/dispatching.c
index 5aecdd1fc..f2c6deada 100644
--- a/numpy/core/src/umath/dispatching.c
+++ b/numpy/core/src/umath/dispatching.c
@@ -1016,8 +1016,13 @@ promote_and_get_ufuncimpl(PyUFuncObject *ufunc,
             signature[i] = (PyArray_DTypeMeta *)PyTuple_GET_ITEM(all_dtypes, i);
             Py_INCREF(signature[i]);
         }
-        else {
-            assert((PyObject *)signature[i] == PyTuple_GET_ITEM(all_dtypes, i));
+        else if ((PyObject *)signature[i] != PyTuple_GET_ITEM(all_dtypes, i)) {
+            /*
+             * If signature is forced we may find an incompatible loop that
+             * would be picked by promotion (signature not forced), reject it.
+             */
+            raise_no_loop_found_error(ufunc, (PyObject **)op_dtypes);
+            return NULL;
         }
     }
```
We still get a slightly different error depending on the code path, but that is because the "normal" error is less precise.

Reply:
OK, maybe not the ideal thing, but I pushed a second commit that applies that fix (with a slightly different comment) and adds a more explicit test.

Reply:
Thank you for fixing this, @seberg!! I was caught up with college and couldn't undo the commit at that time (thanks for fixing that too!)